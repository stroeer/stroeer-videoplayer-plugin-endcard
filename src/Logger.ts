let debugMode = true
if (typeof window !== 'undefined' && typeof window.localStorage !== 'undefined') {
  if (window.localStorage.getItem('StroeerVideoplayerDebugMode') !== null) {
    debugMode = true
  }
}
const logger = {
  log: function (...args: any[]) {
    if (debugMode) {
      console.log.apply(console, args)
    }
  }
}

export default logger