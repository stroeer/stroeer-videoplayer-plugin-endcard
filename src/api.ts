import { IData } from '../types/types'

// https://www.carlrippon.com/fetch-with-async-await-and-typescript/
async function fetchAPI<T> (url: RequestInfo): Promise<T> {
  const response = await fetch(url)
  let body
  if (!response.ok) {
    throw new Error(`HTTP error! status: ${response.status}`)
  }

  try {
    body = await response.json()
  } catch (err) {
    if (typeof err === 'string') {
      console.log(`Something went wrong! ${err}`)
    } else {
      console.log('Something went wrong!')
    }
  }

  return body
}

const transformData: Function = (data: [], keyMap: IData) => {
  for (const [oldKey, newKey] of Object.entries(keyMap)) {
    data.forEach(item => {
      if (!item[oldKey]) return
      Object.assign(item, { [String(newKey)]: item[oldKey] })
      // delete item[oldKey] --> this cause lint error
    })
  }
  return data
}

export {
  fetchAPI,
  transformData
}
