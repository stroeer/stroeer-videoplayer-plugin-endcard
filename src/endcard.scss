$endcard-plugin-highlight-color: #424242 !default;
$endcard-plugin-bg-color: #000 !default;
$endcard-plugin-tile-border: 2px solid #fff !default;
$endcard-plugin-title-bg: linear-gradient(
  180deg,
  rgba(0, 0, 0, 0) 0%,
  rgba(0, 0, 0, 0.8) 100%
) !default;
$endcard-plugin-spacing-large: 16px !default;
$endcard-plugin-spacing-small: 12px !default;
$endcard-plugin-font-size-small: 12px !default;
$endcard-plugin-font-size-smaller: 10px !default;
$endcard-plugin-font-size-medium: 14px !default;
$endcard-plugin-font-size-large: 16px !default;
$endcard-plugin-text-color: #fff !default;
$endcard-plugin-button-pause-bg-color: rgba(255, 255, 255, 0.4) !default;
$endcard-plugin-button-pause-bg-color-hover: rgba(255, 255, 255, 0.2) !default;
$endcard-plugin-tile-overlay-bg: rgba(0, 0, 0, 0.4) !default;
$endcard-plugin-controlbar-height-mobile: 55px !default;

.plugin-endcard-container {
  background-color: $endcard-plugin-bg-color;
  position: absolute;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
}

.plugin-endcard-overlay {
  position: absolute;
  width: 100%;
  bottom: $endcard-plugin-controlbar-height-mobile;
}

.plugin-endcard-tile,
.plugin-endcard-tile-replay {
  display: none;
  position: relative;
  float: left;
  width: 33.33%;
  height: 33.33%;
  box-sizing: border-box;
  cursor: pointer;
  overflow: hidden;
}

.plugin-endcard-tile-replay {
  z-index: 101;
}

.plugin-endcard-tile-single,
.plugin-endcard-tile-0 {
  display: block;
  width: 100%;
  height: 100%;
}

.plugin-endcard-tile-0,
.plugin-endcard-tile-replay {
  &::before {
    content: "";
    position: absolute;
    background-color: $endcard-plugin-tile-overlay-bg;
    width: 100%;
    height: 100%;
  }
}

.plugin-endcard-tile-1,
.plugin-endcard-tile-2 {
  border-bottom: $endcard-plugin-tile-border;
}

.plugin-endcard-tile-replay,
.plugin-endcard-tile-3 {
  border-right: $endcard-plugin-tile-border;
}

.plugin-endcard-tile-single {
  border-right: none;
}

.plugin-endcard-thumbnail {
  width: 100%;
  height: 100%;
  background-repeat: no-repeat;
  background-size: cover;
  background-position: 50%;
  background-color: #000;
}

.plugin-endcard-title-pre {
  display: block;
  margin-bottom: 4px;
  font-weight: normal;
  font-size: $endcard-plugin-font-size-medium;
}

.plugin-endcard-title {
  margin: 0;
  padding: $endcard-plugin-spacing-small;
  width: 100%;
  box-sizing: border-box;
  color: $endcard-plugin-text-color;
  font-weight: bold;
  font-size: $endcard-plugin-font-size-medium;
}

.plugin-endcard-replay-title {
  margin: 0;
  padding-top: $endcard-plugin-spacing-small;
  text-align: center;
  font-size: $endcard-plugin-font-size-smaller;
  color: $endcard-plugin-text-color;
  font-weight: bold;
  text-transform: uppercase;
  letter-spacing: 2px;
}

.plugin-endcard-button-pause {
  position: absolute;
  top: $endcard-plugin-spacing-large;
  right: $endcard-plugin-spacing-large;
  background-color: $endcard-plugin-button-pause-bg-color;
  color: $endcard-plugin-text-color;
  text-transform: uppercase;
  font-weight: 700;
  font-size: $endcard-plugin-font-size-small;
  letter-spacing: 2px;
  padding: 8px;
  border: none;
  border-radius: 4px;
  cursor: pointer;
  z-index: 1;
}

.plugin-endcard-revolverplay-icon {
  display: block;
  margin: 0 auto;
}

.plugin-endcard-play-icon {
  display: block;
  margin: 0 auto $endcard-plugin-spacing-small;
}

.plugin-endcard-replay-icon {
  display: block;
  margin: 0 auto;
}

.plugin-endcard-progress-value {
  stroke: $endcard-plugin-highlight-color;
  stroke-linecap: round;
  transition: all 1s ease;
}

.stroeer-videoplayer .stroeer-videoplayer-ui.plugin-endcard-ui-small {
  height: $endcard-plugin-controlbar-height-mobile;
  top: inherit;
  bottom: 0;

  .default .video-overlay {
    background: none;
  }
}

@media only screen and (min-width: 769px) {
  .plugin-endcard-container {
    z-index: 101;
  }

  .plugin-endcard-tile {
    display: block;
  }

  .plugin-endcard-overlay {
    bottom: 0;
  }

  .plugin-endcard-title {
    background: $endcard-plugin-title-bg;

    .plugin-endcard-tile-0 & {
      font-size: $endcard-plugin-font-size-large;
      padding: $endcard-plugin-spacing-large;
    }
  }

  .plugin-endcard-tile-0 {
    width: 66.66%;
    height: 66.66%;
    border-right: $endcard-plugin-tile-border;
    border-bottom: $endcard-plugin-tile-border;
  }

  .plugin-endcard-tile-single {
    width: 100%;
    height: 100%;
    border-right: none;
    border-bottom: none;
  }

  .plugin-endcard-tile-replay {
    display: block;
    clear: left;

    .plugin-endcard-overlay {
      bottom: calc(50% - ((21px + #{$endcard-plugin-font-size-smaller} + #{$endcard-plugin-spacing-small}) / 2));
    }
  }

  .plugin-endcard-tile-1,
  .plugin-endcard-tile-2,
  .plugin-endcard-tile-3,
  .plugin-endcard-tile-4 {
    .plugin-endcard-overlay {
      transition: all 0.4s;
      opacity: 0;
    }

    &:hover {
      .plugin-endcard-overlay {
        opacity: 1;
      }
    }
  }

  .plugin-endcard-revolverplay-icon {
    margin-bottom: $endcard-plugin-spacing-large;
  }

  .plugin-endcard-button-pause {
    transition: all 0.4s;

    &:hover {
      background-color: $endcard-plugin-button-pause-bg-color-hover;
    }
  }
}
