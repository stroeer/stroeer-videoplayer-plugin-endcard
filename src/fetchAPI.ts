// https://www.carlrippon.com/fetch-with-async-await-and-typescript/
async function fetchAPI<T> (url: RequestInfo): Promise<T> {
  const response = await fetch(url)
  let body
  
  if (!response.ok) {
    throw new Error(`HTTP error! status: ${response.status}`)
  }

  try {
    body = await response.json()
  } catch (err) {
    console.log(`Something went wrong! ${err}`)
  }

  return body
}

export default fetchAPI