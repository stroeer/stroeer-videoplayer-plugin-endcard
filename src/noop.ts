export function noop(): boolean {
  return false
}

export function noopData<T> (data: T): T {
  return data
}
