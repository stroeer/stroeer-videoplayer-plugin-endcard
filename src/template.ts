import { IData } from '../types/types'

const getCard = (index: number, obj: IData, revolverplayTime: number): string => {
  const template = `
    <div class="plugin-endcard-card" data-idx="${index}" data-role="plugin-endcard-card">
      <div class="plugin-endcard-thumbnail" style="background-image: url(${obj.image});"></div>
      <span class="plugin-endcard-title">${obj.title}</span>
      ${index === 0
      ? `
        <div class="plugin-endcard-play-icon">
          <svg class="plugin-endcard-progress-icon" viewBox="0 0 120 120">
            <circle class="plugin-endcard-progress-meter"  cx="60" cy="60" r="54" fill="none" stroke="#e6e6e6" stroke-width="3" />
            <circle class="plugin-endcard-progress-value" cx="60" cy="60" r="54" fill="none" stroke="#f77a52" stroke-width="3" stroke-dasharray="339.292" stroke-dashoffset="339.292" />
            <path fill="#e6e6e6" d="m 60.940678,83.618644 -25,-32 h 51 z"/>
          </svg>
        </div>
        ${revolverplayTime !== 0
          ? '<button class="plugin-endcard-pause" data-role="plugin-endcard-pause">Anhalten</button>'
          : ''}  
      `
      : ''}
    </div>
  `

  return template
}

export default getCard
