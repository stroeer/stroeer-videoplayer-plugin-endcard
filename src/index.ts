import EndcardPlugin from './plugin'
import { version } from '../package.json'
import logger from './logger'
import { IStroeerVideoplayer, IEndcardOptions } from '../types/types'
import './endcard.scss'

const plugin = {
  pluginName: 'Endcard',
  init: (stroeervideoplayer: IStroeerVideoplayer, opts: IEndcardOptions = {}) => {
    logger.log('opts', opts)

    const endcardPlugin = new EndcardPlugin(stroeervideoplayer, opts)
    const videoEl = stroeervideoplayer.getVideoEl()

    videoEl.addEventListener('contentVideoFirstQuartile', () => {
      endcardPlugin.render()
    })

    videoEl.addEventListener('contentVideoEnded', () => {
      endcardPlugin.addClickEvents()
      endcardPlugin.show()
      endcardPlugin.revolverplay()
    })
  },
  deinit: () => {
  },
  version: version
}

export default plugin
