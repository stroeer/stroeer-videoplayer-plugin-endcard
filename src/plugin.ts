import noop from './noop'
import logger from './logger'
import { fetchAPI, transformData } from './api'
import { getTile, getTileReplay } from './template'
import { IData, IStroeerVideoplayer, IEndcardOptions } from '../types/types'
import { ticker } from './revolverplay'

class EndcardPlugin {
  videoplayer: IStroeerVideoplayer
  videoElement: HTMLVideoElement
  endcardContainer: HTMLDivElement
  onLoadedCallback: Function
  onClickToPlayCallback: Function
  onClickToReplayCallback: Function
  onRevolverplayCallback: Function
  onRevolverplayPauseCallback: Function
  dataKeyMap: object
  transformedData: IData[]
  showFallback: boolean
  uiEl: HTMLDivElement
  revolverplayTime: number
  intervalTicker: NodeJS.Timeout | null

  constructor (stroeervideoplayer: IStroeerVideoplayer, opts: IEndcardOptions = {}) {
    this.videoplayer = stroeervideoplayer
    this.videoElement = stroeervideoplayer.getVideoEl()

    this.dataKeyMap = opts.dataKeyMap !== undefined ? opts.dataKeyMap : noop
    this.transformedData = []
    this.showFallback = opts.showFallback !== undefined ? opts.showFallback : false
    this.revolverplayTime = opts.revolverplayTime !== undefined ? opts.revolverplayTime : 5
    this.intervalTicker = null

    this.onLoadedCallback = opts.onLoadedCallback !== undefined ? opts.onLoadedCallback : noop
    this.onClickToPlayCallback = opts.onClickToPlayCallback !== undefined ? opts.onClickToPlayCallback : noop
    this.onClickToReplayCallback = opts.onClickToReplayCallback !== undefined ? opts.onClickToReplayCallback : noop
    this.onRevolverplayCallback = opts.onRevolverplayCallback !== undefined ? opts.onRevolverplayCallback : noop
    this.onRevolverplayPauseCallback = opts.onRevolverplayPauseCallback !== undefined ? opts.onRevolverplayPauseCallback : noop

    this.uiEl = stroeervideoplayer.getUIEl()

    this.endcardContainer = document.createElement('div')
    this.endcardContainer.classList.add('plugin-endcard-container', 'hidden')
    this.videoElement.after(this.endcardContainer)

    return this
  }

  getEndcardUrl = (): string => {
    const url = this.videoElement.dataset.endcardUrl
    return url !== undefined ? url : ''
  }

  setEndcardUrl = (url: string): void => {
    this.videoElement.dataset.endcardUrl = url
  }

  reset = (): void => {
    this.clearRevolverplayTimer()
    this.removeClickEvents()
    this.endcardContainer.innerHTML = ''
    this.hide()
  }

  revolverplay = (): void => {
    if (this.revolverplayTime === 0 || this.showFallback) return

    /* eslint-disable-next-line */
    const progressSvgCircle = this.endcardContainer.querySelector('[data-role="plugin-endcard-progress-value"]')! as HTMLElement
    let remainingTime = this.revolverplayTime
    const revolverplayTicker = (): void => {
      ticker(this.revolverplayTime, remainingTime, progressSvgCircle, () => {
        this.play(0, true)
        this.onRevolverplayCallback()
      })
      remainingTime--
    }

    revolverplayTicker()
    this.intervalTicker = setInterval(revolverplayTicker, 1000)
  }

  clearRevolverplayTimer = (): void => {
    if (this.intervalTicker !== null) {
      clearInterval(this.intervalTicker)
    }
  }

  replay = (): void => {
    this.videoplayer.play()
  }

  play = (idx: number, autoplay: boolean): void => {
    this.setEndcardUrl(this.transformedData[idx].endpoint)
    this.videoplayer.replaceAndPlay(this.transformedData[idx], autoplay)
  }

  clickToPlay = (e: Event): void => {
    e.preventDefault()

    const el = (e.target as Element).closest('[data-role="plugin-endcard-tile"]')
    const idx: string | null = el !== null ? el.getAttribute('data-idx') : null
    if (idx === null) return
    this.play(parseInt(idx), false)
    this.onClickToPlayCallback()
  }

  clickToReplay = (e: Event): void => {
    e.preventDefault()
    e.stopPropagation()
    this.replay()
    this.onClickToReplayCallback()
  }

  clickToPause = (e: Event): void => {
    const circles = this.endcardContainer.querySelectorAll('[data-role="plugin-endcard-revolverplay-icon"] circle')
    const target = e.currentTarget as HTMLElement

    e.preventDefault()
    e.stopPropagation()
    this.onRevolverplayPauseCallback()
    this.clearRevolverplayTimer()

    if (target !== null) {
      target.remove()
    }
    circles.forEach((circle) => {
      circle.remove()
    })
  }

  addClickEvents = (): void => {
    const tiles = this.endcardContainer.querySelectorAll('[data-role="plugin-endcard-tile"]')
    const pauseButton = this.endcardContainer.querySelector('[data-role="plugin-endcard-pause"]')
    const replayTile = this.endcardContainer.querySelector('[data-role="plugin-endcard-tile-replay"]')

    tiles.forEach(tile => {
      tile.addEventListener('click', this.clickToPlay)
    })

    if (replayTile !== null) {
      replayTile.addEventListener('click', this.clickToReplay)
    }

    if (pauseButton !== null) {
      pauseButton.addEventListener('click', this.clickToPause)
    }
  }

  removeClickEvents = (): void => {
    const tiles = this.endcardContainer.querySelectorAll('[data-role="plugin-endcard-tile"]')
    const pauseButton = this.endcardContainer.querySelector('[data-role="plugin-endcard-pause"]')
    const replayTile = this.endcardContainer.querySelector('[data-role="plugin-endcard-tile-replay"]')

    tiles.forEach(tile => {
      tile.removeEventListener('click', this.clickToPlay)
    })

    if (replayTile !== null) {
      replayTile.removeEventListener('click', this.clickToReplay)
    }

    if (pauseButton !== null) {
      pauseButton.removeEventListener('click', this.clickToPause)
    }
  }

  renderFallback = (): void => {
    const replayTemplate = getTileReplay(this.videoplayer.getPosterImage(), 'plugin-endcard-tile-single')
    this.endcardContainer.innerHTML += replayTemplate
  }

  render = (): void => {
    const endpoint = this.getEndcardUrl()
    if (endpoint === null || this.showFallback) {
      this.showFallback = true
      this.renderFallback()
      return
    }

    fetchAPI<object>(endpoint)
      .then((data) => {
        this.transformedData = transformData(data, this.dataKeyMap)
        logger.log(this.transformedData)

        for (let i: number = 0; i < 5; i++) {
          const tileTemplate = getTile(i, this.transformedData[i], this.revolverplayTime)
          const replayTemplate = getTileReplay(this.videoplayer.getPosterImage())
          if (i === 3) {
            this.endcardContainer.innerHTML += replayTemplate
          }
          this.endcardContainer.innerHTML += tileTemplate
        }
      })
      .catch(err => {
        logger.log('Something went wrong with fetching api!', err)
        this.showFallback = true
        this.renderFallback()
      })
  }

  hide = (): void => {
    if (this.uiEl.classList.contains('plugin-endcard-ui-small')) {
      this.uiEl.classList.remove('plugin-endcard-ui-small')
    }
    this.endcardContainer.classList.add('hidden')
  }

  show = (): void => {
    this.uiEl.classList.add('plugin-endcard-ui-small')

    if (typeof this.videoplayer.exitFullscreen === 'function') {
      this.videoplayer.exitFullscreen()
    }
    this.endcardContainer.classList.remove('hidden')
    this.onLoadedCallback()
  }
}

export default EndcardPlugin
