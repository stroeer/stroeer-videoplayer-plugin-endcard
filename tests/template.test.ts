import { getTile, getTileReplay } from '../src/template'

const dataObj = [
  {
    title: 'Title of tile number one',
    image: 'https://videos.giga.de/files/1307753225/480p.jpg',
    sources: [],
    endpoint: '',
  },
  {
    title: 'Title of tile number two',
    image: 'https://videos.giga.de/files/1307753225/720p.jpg',
    sources: [],
    endpoint: '',
  },
  {
    title: 'Title of tile number three',
    image: 'https://videos.giga.de/files/1307753225/1080p.jpg',
    sources: [],
    endpoint: '',
  }
]

test('Tile markup is rendered as expected when revolverplay is active', () => {
  document.body.innerHTML = '<div id="testContainer"></div>'
  const stage = document.querySelector('#testContainer')!
  for (let i = 0; i < dataObj.length; i++) {
    stage.innerHTML += getTile(i, dataObj[i], 2)
  }
  expect(stage).toMatchSnapshot()
})

test('Tile markup is rendered as expected when revolverplay is not active', () => {
  document.body.innerHTML = '<div id="testContainer"></div>'
  const stage = document.querySelector('#testContainer')!
  for (let i = 0; i < dataObj.length; i++) {
    stage.innerHTML += getTile(i, dataObj[i], 0)
  }
  expect(stage).toMatchSnapshot()
})

test('Replay tile markup is rendered as expected', () => {
  document.body.innerHTML = '<div id="testContainer"></div>'
  const stage = document.querySelector('#testContainer')!
  stage.innerHTML = getTileReplay('https://videos.giga.de/files/1307753225/720p.jpg')
  expect(stage).toMatchSnapshot()
})