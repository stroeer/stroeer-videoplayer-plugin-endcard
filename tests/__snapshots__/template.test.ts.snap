// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Replay tile markup is rendered as expected 1`] = `
<div
  id="testContainer"
>
  
    
  <div
    class="plugin-endcard-tile"
    data-role="plugin-endcard-tile-replay"
  >
    
      
    <div
      class="plugin-endcard-thumbnail"
      style="background-image: url(https://videos.giga.de/files/1307753225/720p.jpg);"
    />
    
      
    <div
      class="plugin-endcard-overlay"
    >
      
        
      <svg
        class="plugin-endcard-replay-icon"
        fill="none"
        height="21"
        viewBox="0 0 20 16"
        width="26"
        xmlns="http://www.w3.org/2000/svg"
      >
        
          
        <path
          d="M19 16H1C0.447715 16 0 15.5523 0 15V10C0 9.44772 0.447715 9 1 9C1.55228 9 2 9.44772 2 10V14H18V5H8V8L3 4L8 0V3H19C19.553 3 20 3.447 20 4V15C20 15.553 19.553 16 19 16Z"
          fill="white"
        />
        
        
      </svg>
      
        
      <p
        class="plugin-endcard-replay-title"
      >
        Video wiederholen
      </p>
      
      
    </div>
    
    
  </div>
  
  
</div>
`;

exports[`Tile markup is rendered as expected when revolverplay is active 1`] = `
<div
  id="testContainer"
>
  
    
  <div
    class="plugin-endcard-tile"
    data-idx="0"
    data-role="plugin-endcard-tile"
  >
    
      
    <button
      class="plugin-endcard-button-pause"
      data-role="plugin-endcard-pause"
    >
      Anhalten
    </button>
      
      
    <div
      class="plugin-endcard-thumbnail"
      style="background-image: url(https://videos.giga.de/files/1307753225/480p.jpg);"
    />
    
      
    <div
      class="plugin-endcard-overlay"
      data-role="plugin-endcard-overlay"
    >
      
        
          
      <svg
        class="plugin-endcard-revolverplay-icon"
        fill="none"
        height="75"
        width="103"
        xmlns="http://www.w3.org/2000/svg"
      >
        
            
        <g
          filter="url(#filter0_d)"
        >
          
              
          <path
            d="M62.073 37.89l-20-13.333A1.334 1.334 0 0040 25.667v26.666a1.334 1.334 0 002.073 1.11l20-13.334a1.335 1.335 0 000-2.218z"
            fill="#fff"
          />
          
            
        </g>
        
            
        <circle
          class="plugin-endcard-progress-meter"
          cx="48"
          cy="39"
          fill="none"
          r="30.667"
          stroke="#fff"
          stroke-width="2.667"
        />
        
            
        <circle
          class="plugin-endcard-progress-value"
          cx="48"
          cy="39"
          data-role="plugin-endcard-progress-value"
          fill="none"
          r="30.667"
          stroke="#fff"
          stroke-dasharray="192.686"
          stroke-dashoffset="192.686"
          stroke-width="2.667"
          transform="rotate(-90 48 39)"
        />
        
            
        <defs>
          
              
          <filter
            color-interpolation-filters="sRGB"
            filterUnits="userSpaceOnUse"
            height="112"
            id="filter0_d"
            width="112"
            x="-5.333"
            y="-1"
          >
            
                
            <feflood
              flood-opacity="0"
              result="BackgroundImageFix"
            />
            
                
            <fecolormatrix
              in="SourceAlpha"
              values="0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 127 0"
            />
            
                
            <feoffset
              dy="16"
            />
            <fegaussianblur
              stdDeviation="20"
            />
            
                
            <fecolormatrix
              values="0 0 0 0 0.490196 0 0 0 0 0.596078 0 0 0 0 0.698039 0 0 0 0.2 0"
            />
            
                
            <feblend
              in2="BackgroundImageFix"
              mode="multiply"
              result="effect1_dropShadow"
            />
            
                
            <feblend
              in="SourceGraphic"
              in2="effect1_dropShadow"
              result="shape"
            />
            
              
          </filter>
          
            
        </defs>
        
          
      </svg>
       
        
        
      <p
        class="plugin-endcard-title"
      >
        
          
        <span
          class="plugin-endcard-title-pre"
        >
          NÃ¤chstes Video
        </span>
        
          Title of tile number one
        
      </p>
      
      
    </div>
    
    
  </div>
  

  
    
  <div
    class="plugin-endcard-tile"
    data-idx="1"
    data-role="plugin-endcard-tile"
  >
    
        
      
    <div
      class="plugin-endcard-thumbnail"
      style="background-image: url(https://videos.giga.de/files/1307753225/720p.jpg);"
    />
    
      
    <div
      class="plugin-endcard-overlay"
      data-role="plugin-endcard-overlay"
    >
      
        
        
      <svg
        class="plugin-endcard-play-icon"
        fill="none"
        height="22"
        viewBox="0 0 17 22"
        width="17"
        xmlns="http://www.w3.org/2000/svg"
      >
        
          
        <path
          d="M16.555 10.168L1.555 0.168C1.248 -0.0359995 0.853 -0.0559995 0.529 0.118001C0.203 0.292001 0 0.631001 0 1V21C0 21.369 0.203 21.708 0.528 21.882C0.676 21.961 0.838 22 1 22C1.194 22 1.388 21.943 1.555 21.832L16.555 11.832C16.833 11.646 17 11.334 17 11C17 10.666 16.833 10.354 16.555 10.168Z"
          fill="white"
        />
        
        
      </svg>
      
        
        
      <p
        class="plugin-endcard-title"
      >
        
          
          Title of tile number two
        
      </p>
      
      
    </div>
    
    
  </div>
  

  
    
  <div
    class="plugin-endcard-tile"
    data-idx="2"
    data-role="plugin-endcard-tile"
  >
    
        
      
    <div
      class="plugin-endcard-thumbnail"
      style="background-image: url(https://videos.giga.de/files/1307753225/1080p.jpg);"
    />
    
      
    <div
      class="plugin-endcard-overlay"
      data-role="plugin-endcard-overlay"
    >
      
        
        
      <svg
        class="plugin-endcard-play-icon"
        fill="none"
        height="22"
        viewBox="0 0 17 22"
        width="17"
        xmlns="http://www.w3.org/2000/svg"
      >
        
          
        <path
          d="M16.555 10.168L1.555 0.168C1.248 -0.0359995 0.853 -0.0559995 0.529 0.118001C0.203 0.292001 0 0.631001 0 1V21C0 21.369 0.203 21.708 0.528 21.882C0.676 21.961 0.838 22 1 22C1.194 22 1.388 21.943 1.555 21.832L16.555 11.832C16.833 11.646 17 11.334 17 11C17 10.666 16.833 10.354 16.555 10.168Z"
          fill="white"
        />
        
        
      </svg>
      
        
        
      <p
        class="plugin-endcard-title"
      >
        
          
          Title of tile number three
        
      </p>
      
      
    </div>
    
    
  </div>
  

  
</div>
`;

exports[`Tile markup is rendered as expected when revolverplay is not active 1`] = `
<div
  id="testContainer"
>
  
    
  <div
    class="plugin-endcard-tile"
    data-idx="0"
    data-role="plugin-endcard-tile"
  >
    
        
      
    <div
      class="plugin-endcard-thumbnail"
      style="background-image: url(https://videos.giga.de/files/1307753225/480p.jpg);"
    />
    
      
    <div
      class="plugin-endcard-overlay"
      data-role="plugin-endcard-overlay"
    >
      
        
          
      <svg
        class="plugin-endcard-revolverplay-icon"
        fill="none"
        height="75"
        width="103"
        xmlns="http://www.w3.org/2000/svg"
      >
        
            
        <g
          filter="url(#filter0_d)"
        >
          
              
          <path
            d="M62.073 37.89l-20-13.333A1.334 1.334 0 0040 25.667v26.666a1.334 1.334 0 002.073 1.11l20-13.334a1.335 1.335 0 000-2.218z"
            fill="#fff"
          />
          
            
        </g>
        
            
        <circle
          class="plugin-endcard-progress-meter"
          cx="48"
          cy="39"
          fill="none"
          r="30.667"
          stroke="#fff"
          stroke-width="2.667"
        />
        
            
        <circle
          class="plugin-endcard-progress-value"
          cx="48"
          cy="39"
          data-role="plugin-endcard-progress-value"
          fill="none"
          r="30.667"
          stroke="#fff"
          stroke-dasharray="192.686"
          stroke-dashoffset="192.686"
          stroke-width="2.667"
          transform="rotate(-90 48 39)"
        />
        
            
        <defs>
          
              
          <filter
            color-interpolation-filters="sRGB"
            filterUnits="userSpaceOnUse"
            height="112"
            id="filter0_d"
            width="112"
            x="-5.333"
            y="-1"
          >
            
                
            <feflood
              flood-opacity="0"
              result="BackgroundImageFix"
            />
            
                
            <fecolormatrix
              in="SourceAlpha"
              values="0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 127 0"
            />
            
                
            <feoffset
              dy="16"
            />
            <fegaussianblur
              stdDeviation="20"
            />
            
                
            <fecolormatrix
              values="0 0 0 0 0.490196 0 0 0 0 0.596078 0 0 0 0 0.698039 0 0 0 0.2 0"
            />
            
                
            <feblend
              in2="BackgroundImageFix"
              mode="multiply"
              result="effect1_dropShadow"
            />
            
                
            <feblend
              in="SourceGraphic"
              in2="effect1_dropShadow"
              result="shape"
            />
            
              
          </filter>
          
            
        </defs>
        
          
      </svg>
       
        
        
      <p
        class="plugin-endcard-title"
      >
        
          
        <span
          class="plugin-endcard-title-pre"
        >
          NÃ¤chstes Video
        </span>
        
          Title of tile number one
        
      </p>
      
      
    </div>
    
    
  </div>
  

  
    
  <div
    class="plugin-endcard-tile"
    data-idx="1"
    data-role="plugin-endcard-tile"
  >
    
        
      
    <div
      class="plugin-endcard-thumbnail"
      style="background-image: url(https://videos.giga.de/files/1307753225/720p.jpg);"
    />
    
      
    <div
      class="plugin-endcard-overlay"
      data-role="plugin-endcard-overlay"
    >
      
        
        
      <svg
        class="plugin-endcard-play-icon"
        fill="none"
        height="22"
        viewBox="0 0 17 22"
        width="17"
        xmlns="http://www.w3.org/2000/svg"
      >
        
          
        <path
          d="M16.555 10.168L1.555 0.168C1.248 -0.0359995 0.853 -0.0559995 0.529 0.118001C0.203 0.292001 0 0.631001 0 1V21C0 21.369 0.203 21.708 0.528 21.882C0.676 21.961 0.838 22 1 22C1.194 22 1.388 21.943 1.555 21.832L16.555 11.832C16.833 11.646 17 11.334 17 11C17 10.666 16.833 10.354 16.555 10.168Z"
          fill="white"
        />
        
        
      </svg>
      
        
        
      <p
        class="plugin-endcard-title"
      >
        
          
          Title of tile number two
        
      </p>
      
      
    </div>
    
    
  </div>
  

  
    
  <div
    class="plugin-endcard-tile"
    data-idx="2"
    data-role="plugin-endcard-tile"
  >
    
        
      
    <div
      class="plugin-endcard-thumbnail"
      style="background-image: url(https://videos.giga.de/files/1307753225/1080p.jpg);"
    />
    
      
    <div
      class="plugin-endcard-overlay"
      data-role="plugin-endcard-overlay"
    >
      
        
        
      <svg
        class="plugin-endcard-play-icon"
        fill="none"
        height="22"
        viewBox="0 0 17 22"
        width="17"
        xmlns="http://www.w3.org/2000/svg"
      >
        
          
        <path
          d="M16.555 10.168L1.555 0.168C1.248 -0.0359995 0.853 -0.0559995 0.529 0.118001C0.203 0.292001 0 0.631001 0 1V21C0 21.369 0.203 21.708 0.528 21.882C0.676 21.961 0.838 22 1 22C1.194 22 1.388 21.943 1.555 21.832L16.555 11.832C16.833 11.646 17 11.334 17 11C17 10.666 16.833 10.354 16.555 10.168Z"
          fill="white"
        />
        
        
      </svg>
      
        
        
      <p
        class="plugin-endcard-title"
      >
        
          
          Title of tile number three
        
      </p>
      
      
    </div>
    
    
  </div>
  

  
</div>
`;
